src/main/java/com/amazon/opendistroforelasticsearch/security/auditlog/impl/AuditLogImpl.java

```diff
@@ -38,179 +53,468 @@
+    private final String opendistrosecurityIndex;
+
+    private static final List<String> writeClasses = Arrays.asList(
+            IndexRequest.class.getSimpleName(),
+            UpdateRequest.class.getSimpleName(),
+            BulkRequest.class.getSimpleName(),
+            BulkShardRequest.class.getSimpleName(),
+            DeleteRequest.class.getSimpleName()
+    );
+
+    public AuditLogImpl(final Settings settings,
+                        final Path configPath,
+                        final Client clientProvider,
+                        final ThreadPool threadPool,
+                        final IndexNameExpressionResolver indexNameExpressionResolver,
+                        final ClusterService clusterService) {
+        this.threadPool = threadPool;
+        this.settings = settings;
+        this.clusterService = clusterService;
+        this.opendistrosecurityIndex = settings.get(ConfigConstants.OPENDISTRO_SECURITY_CONFIG_INDEX_NAME, ConfigConstants.OPENDISTRO_SECURITY_DEFAULT_CONFIG_INDEX);
+        this.auditConfig = AuditConfig.getConfig(settings);
+        this.requestResolver = new RequestResolver(clusterService, indexNameExpressionResolver, opendistrosecurityIndex, threadPool);
+        this.complianceResolver = new ComplianceResolver(clusterService, indexNameExpressionResolver, opendistrosecurityIndex);
+        this.messageRouter = new AuditMessageRouter(settings, clientProvider, threadPool, configPath);
+        this.enabled = messageRouter.isEnabled();
+
+        log.info("Message routing enabled: {}", this.enabled);
+
+        final SecurityManager sm = System.getSecurityManager();
+
+        if (sm != null) {
+            log.debug("Security Manager present");
+            sm.checkPermission(new SpecialPermission());
+        }
+
+        AccessController.doPrivileged((PrivilegedAction<Object>) () -> {
+            Runtime.getRuntime().addShutdownHook(new Thread(() -> {
+                try {
+                    close();
+                } catch (final IOException e) {
+                    log.warn("Exception while shutting down message router", e);
+                }
+            }));
+            log.debug("Shutdown Hook registered");
+            return null;
+        });
+    }
+
+    @Override
+    public void setComplianceConfig(final ComplianceConfig complianceConfig) {
+        messageRouter.setComplianceConfig(complianceConfig);
+    }
+
+    @Override
+    public void close() throws IOException {
+        messageRouter.close();
+    }
+
+    protected void save(final AuditMessage msg) {
+        if (enabled) {
+            messageRouter.route(msg);
+        }
+    }
+
+    @Override
+    public void logFailedLogin(final String effectiveUser,
+                               final boolean securityAdmin,
+                               final String initiatingUser,
+                               final TransportRequest request,
+                               final Task task) {
+        if (!enabled) return;
```

Comment:
since save already checks for enabled is there any advantage of having the enabled check for each audit log?